### ISSUE TITLE
Support compiling with clang and -stdlib=libc++

### ISSUE THREAD

**Issue by brunner6** [LLNL/Umpire#321]
**Describe the bug**

There is some code to demangle C++ names in tools/replay/ReplayInterpreter.cpp that relies on a non-standard GCC extension, `abi::__cxa_demangle`.  When compiling with clang and using the -stdlib=libc++ option, which uses LLVM's implementation of the standard library, this function is not available.

**To Reproduce**

Compile the code with
```
module load clang/9.0.0
cmake  ../Umpire -DCMAKE_BUILD_TYPE=Release  -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_FLAGS="-stdlib=libc++ -DGTEST_HAS_CXXABI_H_=0" -DCMAKE_C_FLAGS="-DGTEST_HAS_CXXABI_H_=0" -DENABLE_CUDA=OFF -DENABLE_FORTRAN=OFF -DENABLE_C=ON  -DCMAKE_BUILD_TYPE=Release -DENABLE_TESTS=TRUE -DENABLE_GTEST=TRUE -DENABLE_GMOCK=TRUE -DENABLE_FRUIT=FALSE -DENABLE_EXAMPLES=TRUE -DENABLE_BENCHMARKS=TRUE -DCMAKE_VERBOSE_MAKEFILE=TRUE -DBLT_SOURCE_DIR=../blt -DENABLE_TOOLS=ON
```

**Expected behavior**

The code compiles with clang and libc++

**Compilers & Libraries (please complete the following information):**
 - clang (any version) with the -stdlib=libcxx flag
 - you also need to Define `-DGTEST_HAS_CXXABI_H_=0` in the C/C++ flags to get around the same exact issue in google test.

**Additional context**

This compiler set/option is one that KULL uses regularly to test.

I have a branch with with this patch on it, but I can't push to the repo:
```
diff --git a/tools/replay/ReplayInterpreter.cpp b/tools/replay/ReplayInterpreter.cpp
index d0e5859b..b8a9e383 100644
--- a/tools/replay/ReplayInterpreter.cpp
+++ b/tools/replay/ReplayInterpreter.cpp
@@ -16,7 +16,8 @@
 #include "ReplayFile.hpp"
 #include "umpire/tpl/json/json.hpp"
 
-#if !defined(_MSC_VER)
+#include <ciso646>
+#if !defined(_MSC_VER) && !defined(_LIBCPP_VERSION)
 #include <cxxabi.h>
 #endif
 
@@ -352,6 +353,9 @@ void ReplayInterpreter::replay_compileAllocator( void )
       const std::string mangled_type = 
         (type_prefix == "_Z") ? raw_mangled_type : std::string{"_Z"} + raw_mangled_type;
 
+#if defined(_MSC_VER) || defined(_LIBCPP_VERSION)
+      type = mangled_type;
+#else
       auto result = abi::__cxa_demangle(
           mangled_type.c_str(),
           nullptr,
@@ -362,6 +366,7 @@ void ReplayInterpreter::replay_compileAllocator( void )
       }
       type = std::string{result};
       ::free(result);
+#endif
     } else {
       type = raw_mangled_type;
     }
```

**Comment by mcfadden8** [comment]
Thanks @brunner6, since the replay tool depends on the demangled signatures in order to reconstruct (replay the makeAllocator calls), I guess that we should address this by updating the CMakeList file to not include the replay toolset if their is no support for demangling.

Are there other Umpire tools you are interested in using with with `-stdlib=libc++`?  I'm want to make sure I understand why you want to build with `ENABLE_TOOLS=ON`.

Marty

**Comment by brunner6** [comment]
At one point this code wasn't protected by the `ENABLE_TOOLS=ON`.  The last version of Umpire we had, we needed to apply this patch in order to build.  (Or a similar one anyway.)

In general, we like testing with "odd" compilers so that you can find non-standard usage of the library, and calling `abi::__cxa_demangle` counts.  (Microsoft doesn't support it either, and this tool is broken there too.)

Does the tool depend on the demangled signatures, or just the signatures?  I think my patch was just trying to use the mangled name, which would be consistent for record and playback, wouldn't it?

**Comment by mcfadden8** [comment]
I think it is necessary for `replay` to be able to generate template arguments, specifically, the strategy type.  For example, when `replay`  is run, it will parse the following from the replay input file:
```
{ "kind":"replay", "uid":146816, "timestamp":1580324033707901706, "event": "makeAllocator", "payload": { "type":"N6umpire8strategy9MixedPoolE", "with_introspection":true, "allocator_name":"HOST_MixedPool_spec_0", "args": [ "HOST" ] } }
```
and will generate the following call from Umpire:
```
auto& rm = umpire::ResourceManager::getInstance();
rm.makeAllocator<umpire::strategy::MixedPool, true>(name, args...);
```
So the implementation needs to demangle `"type":"N6umpire8strategy9MixedPoolE"` to `umpire::strategy::MixedPool`.

**Comment by mcfadden8** [comment]
I am placing appropriate guards around the demangling calls as part of the backtrace stuff I am working on.  I will link this issue to an upcoming PR and we will get this fixed.

**Comment by mcfadden8** [comment]
This has been fixed in the 'develop' branch and will be made available with the 2.2.0 release of Umpire
